modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/database/H2TreeInlineObjectDetector.java

```diff
@@ -68,26 +85,61 @@
                 continue;
             }
 
-            if (r.getValue(ih.columnIndex()) == ValueNull.INSTANCE)
+            Value val = r.getValue(ih.columnIndex());
+
+            if (val == ValueNull.INSTANCE)
                 return false;
 
             int type = PageUtils.getByte(pageAddr, off + fieldOff);
 
-            if (type == 0) {
-                inlineObjSupported = false;
+            //We can have garbage in memory and need to compare data.
+            if (type == Value.JAVA_OBJECT) {
+                int len = PageUtils.getShort(pageAddr, off + fieldOff + 1);
 
-                return true;
-            }
-            else if (type == Value.JAVA_OBJECT) {
-                inlineObjSupported = true;
+                boolean isFull = (len & 0x8000) == 0;
+
+                len = len & 0x7FFF;
+
+                byte[] originalObjBytes = val.getBytesNoCopy();
+
+                // read size more then available space or more then origin length
+                if(len > inlineSize - fieldOff - 3 || len > originalObjBytes.length){
```

Comment:
`len > inlineSize - fieldOff - 3` should not lead to decision that `JAVA_OBJECT` is not supported, it means that there is not enough space to inline for current entry.

Reply:
len - it's inlined length, not length of real object, so it should works.
we just check that inlined len fit for remain space, if not - we can decide that it was garbage